src/freenet/clients/http/ToadletContextImpl.java

```diff
@@ -376,15 +376,32 @@ static void sendReplyHeaders(OutputStream sockOutputStream, int replyCode, Strin
 			}
 		if(contentLength >= 0)
 			mvt.put("content-length", Long.toString(contentLength));
-		
-		String expiresTime;
+
+		boolean allowCaching; // For privacy reasons, only static
+							  // content may be cached
 		if (mTime == null) {
-			expiresTime = "Thu, 01 Jan 1970 00:00:00 GMT";
+			allowCaching = false;
+		} else {
+			allowCaching = true;
+		}
+		String expiresTime;
+		String cacheControl;
+		if (allowCaching) {
+			// use an expiry time of 30 day from now, about the frequency of Freenet releases
+			// Expires is needed for older browsers
+			expiresTime = TimeUtil.makeHTTPDate(System.currentTimeMillis() + DAYS.toMillis(30));
+			cacheControl = "public, max-age=" + String.valueOf(3600 * 24 * 30);
+			// Use contentlength as simplest possible hash function
+			// which needs no additional file access
+			mvt.put("ETag", String.valueOf(contentLength));
```

Comment:
@Bombe do you know off-hand how to get the Freenet version string? That’s something I’d love to use here. If there isn’t a version string, it’s a development version.

Can we access the checksum of the Freenet jar file?

(for the sake of the argument: There are also unlimited changes you can perform to a file which do change its length — many more than the changes you can perform which do not change the length, so the chance of hitting one which does not change the length is pretty slim. Algebra of infinites is fun ☺)